### PR TITLE
Feature/backward

### PR DESCRIPTION
**Pull request by reyoung**

### CODE REVIEW

File: paddle/framework/backward.cc [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+   * `no_grad_names` set. Return an EmptyOp.
+   */
+  if (AllInSet(forwardOp.outputs_, OperatorBase::GRAD_VAR_SUFFIX(),
+               no_grad_names)) {
+    for (auto& name : forwardOp.inputs_) {
+      /// Mark all input is not need
+      no_grad_names.insert(name + OperatorBase::GRAD_VAR_SUFFIX());
+    }
+    return EmptyOp();
+  }
```
Comment (qingqing01) [inline paddle/framework/backward.cc]: 我可以理解所有输入都不需要计算梯度时，可标记所有输出都不计算梯度。
但是，没有太想明白，什么情况下，依据输出不需要计算梯度，来标记输入不计算梯度。

File: paddle/framework/backward.cc [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+
+  //! Returned gradient network
+  auto net = std::make_shared<NetOp>();
+
+  if (forwardOp.IsNetOp()) {
```
Comment (qingqing01) [inline paddle/framework/backward.cc]: 为什么不能把获取NetOp的Backward Op写在NetOp对应的一个方法里？
会遇到其他复杂Op(SwitchOp？ 或许举得例子不合适)还需要再写一个分支的情况吗？

File: paddle/framework/backward.cc [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+  return net_op;
+}
+
+/**
+ * @brief Backward an operator, implementation
+ * @param forwardOp the forward operator
+ * @param no_grad_names variable names not calculate for gradient. Like X@GRAD
+ * is not needed.
+ * @param uniq_id a unique index used inside BackwardImpl, it will be shared
+ * through recursive invoke.
```
Comment (qingqing01) [inline paddle/framework/backward.cc]: Need more comments for `uniq_id`.

File: paddle/framework/backward.cc [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+  } else {
+    //! TODO(fjy)
+    std::shared_ptr<OperatorBase> grad_op = OpRegistry::CreateGradOp(forwardOp);
+    for (std::string& grad_input : grad_op->inputs_) {
+      if (no_grad_names.count(grad_input)) {
+        std::string prefix = grad_input.substr(
+            0, grad_input.size() - OperatorBase::GRAD_VAR_SUFFIX().size());
+        grad_input = prefix + OperatorBase::ZERO_VAR_SUFFIX();
+        net->AddOp(OpRegistry::CreateOp("fill_zeros_like", {prefix},
+                                        {grad_input}, {}));
```
Comment (qingqing01) [inline paddle/framework/backward.cc]: need more comments for  `fill_zeros_like` op.

File: paddle/framework/backward.cc (hunk 2 above)
Comment (dzhwinter) [inline paddle/framework/backward.cc]: @qingqing01 我理解这里是Backward的生成逻辑，和NetOp自身的反向不在同一个层面。Gradient Operator是可插拔Unit，Backward是系统core的一部分。

File: paddle/framework/backward.md [hunk 5]
```diff
@@ -0,0 +1,38 @@
+## Operator/expression 's Backward
+
+### Motivation
+
+In Neural Network, the backpropagation algorithm follows the chain rule, so we need to compound the fundmental gradient operators/expressions together with chain rule . Every forward network need a backward network to construct the full computation lineage, the operator/ expression's Backward feature will generate the backward pass respect to forward pass.
```
Comment (jacquesqiao) [inline paddle/framework/backward.md]: computation lineage  ==》 computation graph？ I can find any definition about computation lineage

File: paddle/framework/backward.md [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+### Implement : gradient operator registry
+
+|                        | forward operator | backward operator                |
+| ---------------------- | ---------------- | -------------------------------- |
+| **Operator::inputs_**  | Inputs           | Inputs, Outputs, OutputGradients |
+| **Operator::outputs_** | Outputs          | InputGradients                   |
+
+Inputs/Outputs means the input/output of the operator,  InputGradients/OutputGradients is the gradient respect to forward opeartor. Forward operator and Backward operator are isomorphic, save their corresponding needs into member attribute.
+
+We use a global hash map record the gradient operators available, follow the philosophy  of minimum core, make operator pluggable unit. Each gradient is an operator and it needs to regist itself.
```
Comment (jacquesqiao) [inline paddle/framework/backward.md]: I think we need not emphasize that we use `hash map`, `map` is enough, `hash map` is a way of optimization.

Comment (jacquesqiao) [inline paddle/framework/backward.md]: regist  ==> register

File: paddle/framework/backward.md [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+
+given a forward network, it generates the backward network. We only care about the Gradients—`OutputGradients`,`InputGradients`.
+
+1. bla bla bla (yuyang)
+
+2. NetOp 
+
+   when the input forward network is a NetOp, it need to call the sub NetOp/Operators backward function recursively and ensure them done. During the process, we need to collect the `OutputGradients` name.
+
+   We share variable in the same scope, as a result, duplicate operator `OutputGradients` will overwirte then duplicate variable.
```
Comment (jacquesqiao) [inline paddle/framework/backward.md]: overwirte  => overwrite, then  => the
I remember that we will add a `add_op` if some outputs are duplicated, and rename the duplicated ones.

File: paddle/framework/backward.cc [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+  ret_val.reserve(names.size());
+  for (size_t i = 0; i < names.size(); ++i) {
+    if (set.find(names[i] + suffix) != set.end()) {
+      ret_val.push_back(i);
+    }
+  }
+  return ret_val;
+}
+
+static std::shared_ptr<OperatorBase> EmptyOp() {
```
Comment (jacquesqiao) [inline paddle/framework/backward.cc]: should we move this function to op registry?

**Comment by wangkuiyi** [comment]
https://stackoverflow.com/questions/9243646/recursive-functions-two-functions-or-last-optional-parameter/9243666#9243666  BackwardImpl => BackwardRecursive

File: paddle/operators/fill_zeros_like_op.h [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+#include "paddle/framework/operator.h"
+
+namespace paddle {
+namespace operators {
+
+template <typename Place, typename T>
+class FillZerosLikeKernel : public framework::OpKernel {
+public:
+  void Compute(const framework::KernelContext& context) const override {
+    auto* output = context.Output(0)->GetMutable<framework::Tensor>();
```
Comment (jacquesqiao) [inline paddle/operators/fill_zeros_like_op.h]: auto output

File: paddle/operators/fill_zeros_like_op.cu [hunk 10]
```diff
@@ -0,0 +1,6 @@
+#include "paddle/framework/op_registry.h"
+#include "paddle/operators/fill_zeros_like_op.h"
+
+REGISTER_OP_GPU_KERNEL(
+    fill_zeros_like,
```
Comment (jacquesqiao) [inline paddle/operators/fill_zeros_like_op.cu]: why name this `fill_zerors_like`, numpy call it [zeros_like](https://docs.scipy.org/doc/numpy/reference/generated/numpy.zeros_like.html)

File: paddle/operators/fill_zeros_like_op.h [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+#include "glog/logging.h"
+#include "paddle/framework/eigen.h"
+#include "paddle/framework/operator.h"
+
+namespace paddle {
+namespace operators {
+
+template <typename Place, typename T>
+class FillZerosLikeKernel : public framework::OpKernel {
+public:
```
Comment (jacquesqiao) [inline paddle/operators/fill_zeros_like_op.h]: please add a Python unit test for this operator

File: paddle/framework/operator.cc [hunk 12]
```diff
@@ -34,46 +34,52 @@ KernelContext::GetEigenDevice<platform::GPUPlace, Eigen::GpuDevice>() const {
...
+    return outputs_.at((size_t)idx);
   }
 }
 
 std::vector<std::string> OperatorBase::Outputs(const std::string& name) const {
+  PADDLE_ENFORCE(in_out_idxs_ != nullptr, "InOut Indice could not be nullptr");
   auto output_format = GetAttr<std::vector<int>>("output_format");
   auto offset = in_out_idxs_->at(name);
-
+  PADDLE_ENFORCE(output_format.at((size_t)offset + 1) <= (int)outputs_.size(),
```
Comment (jacquesqiao) [inline paddle/framework/operator.cc]: (int)  ==> static_cast<int>

File: paddle/framework/net_op_test.cc [hunk 13]
```diff
@@ -72,20 +74,17 @@ TEST(OpKernel, all) {
...
-//    op->DebugString();
-//  }
-//}
+TEST(Net, insert_op) {
+  NetOp net;
+  auto op1 = std::make_shared<EmptyOp>();
+  op1->inputs_ = {"x", "w1", "b1"};
+  op1->outputs_ = {"y"};
+  net.AddOp(op1);
+  net.InsertOp(0, op1);
```
Comment (jacquesqiao) [inline paddle/framework/net_op_test.cc]: check if the inserted op is in the right place

File: paddle/framework/backward.cc [hunk 14]
```diff
@@ -0,0 +1,190 @@
...
+ * @param no_grad_names variable names not calculate for gradient. Like X@GRAD
+ * is not needed.
+ * @param uniq_id a unique index used inside BackwardImpl, it will be shared
+ * through recursive invoke.
+ * @return The backward operator. For simple situation, it is a simple operator.
+ * For complex situation, it is a NetOp.
+ *
+ * See Backward.h for details
+ */
+static std::shared_ptr<OperatorBase> BackwardImpl(
```
Comment (jacquesqiao) [inline paddle/framework/backward.cc]: Is MakeBackwardOp or GetBackwardOp a better name?

File: paddle/framework/backward.cc [hunk 15]
```diff
@@ -0,0 +1,190 @@
...
+   *  just return an EmptyOp. Not return null ptr because EmptyOp does not take
+   *  too much time for calculation, but it is useful for simplifying logic.
+   */
+  if (AllInSet(forwardOp.inputs_, OperatorBase::GRAD_VAR_SUFFIX(),
+               no_grad_names)) {
+    return EmptyOp();
+  }
+
+  /**
+   * All output gradients of forwarding operator do not need to calculate. Then
```
Comment (jacquesqiao) [inline paddle/framework/backward.cc]: add `If`

File: paddle/framework/backward.cc [hunk 16]
```diff
@@ -0,0 +1,190 @@
...
+        dup_outputs.push_back(name + "@RENAME@" + std::to_string(uid) + "@" +
+                              std::to_string(i));
+        net->ops_[op_offset]->Rename(name, dup_outputs.back());
+      }
+      insert_position.push_back(
+          {dup_op.back(),
+           OpRegistry::CreateOp(
+               "add", {dup_outputs}, {name},
+               {{"input_format",
+                 std::vector<int>{0, (int)dup_outputs.size()}}})});
```
Comment (jacquesqiao) [inline paddle/framework/backward.cc]: static_cast<int>

File: paddle/framework/backward.cc (hunk 14 above)
Comment (jacquesqiao) [inline paddle/framework/backward.cc]: This function is not in a class, why need `static`?

File: paddle/framework/backward.cc [hunk 17]
```diff
@@ -0,0 +1,190 @@
...
+  //! Returned gradient network
+  auto net = std::make_shared<NetOp>();
+
+  if (forwardOp.IsNetOp()) {
+    /// Because forwardOp is a net op, it can static_cast.
+    auto& forwardNet = static_cast<const NetOp&>(forwardOp);
+
+    //! Map from output gradient variable name to operator's indices in backward
+    //! net. That operator generates that variable.
+    std::unordered_map<std::string, std::vector<size_t>> dup_output_ops;
```
Comment (jacquesqiao) [inline paddle/framework/backward.cc]: what is `dup` mean？

File: paddle/framework/backward.cc [hunk 18]
```diff
@@ -0,0 +1,190 @@
...
+
+#include "paddle/framework/backward.h"
+#include <list>
+#include "paddle/framework/net.h"
+#include "paddle/framework/op_registry.h"
+
+namespace paddle {
+namespace framework {
+
+static bool AllInSet(const std::vector<std::string>& names,
```
Comment (jacquesqiao) [inline paddle/framework/backward.cc]: Why do we need to use `static`?

File: paddle/framework/backward.cc (hunk 1 above)
Comment (reyoung) [inline paddle/framework/backward.cc]: 在反向遍历的时候，如果使用输出Variable的Operator都没有计算梯度，自己也没办法计算梯度。

File: paddle/framework/backward.cc (hunk 2 above)
Comment (reyoung) [inline paddle/framework/backward.cc]: > 会遇到其他复杂Op(SwitchOp？ 或许举得例子不合适)还需要再写一个分支的情况吗?

或许复杂Op的backward我们能注册到另一个地方。但是这个注册机制感觉非常不统一。。最简单的情况肯定是这么写，在没想到更好的办法的时候，先这么做吧。

File: paddle/framework/backward.cc [hunk 19]
```diff
@@ -0,0 +1,190 @@
...
+      for (size_t i = 0; i < dup_op.size(); ++i) {
+        auto op_offset = dup_op[i];
+        dup_outputs.push_back(name + "@RENAME@" + std::to_string(uid) + "@" +
+                              std::to_string(i));
+        net->ops_[op_offset]->Rename(name, dup_outputs.back());
+      }
+      insert_position.push_back(
+          {dup_op.back(),
+           OpRegistry::CreateOp(
+               "add", {dup_outputs}, {name},
```
Comment (jacquesqiao) [inline paddle/framework/backward.cc]: 这个add op现在应该还没实现？

File: paddle/framework/backward_test.cc [hunk 20]
```diff
@@ -0,0 +1,388 @@
...
+  auto gop = f::OpRegistry::CreateGradOp(*fwd);
+  ASSERT_EQ(1UL, gop->inputs_.size());
+  ASSERT_EQ("Out" + f::OperatorBase::GRAD_VAR_SUFFIX(), gop->inputs_[0]);
+  ASSERT_EQ("rowwise_add_grad", gop->type_);
+  ASSERT_EQ("X" + f::OperatorBase::GRAD_VAR_SUFFIX(), gop->outputs_[0]);
+  ASSERT_EQ("b" + f::OperatorBase::GRAD_VAR_SUFFIX(), gop->outputs_[1]);
+
+  ASSERT_EQ("X" + f::OperatorBase::GRAD_VAR_SUFFIX(),
+            gop->Output("X" + f::OperatorBase::GRAD_VAR_SUFFIX()));
+  //  LOG(INFO) << gop->Output("X" + "@GRAD");
```
Comment (jacquesqiao) [inline paddle/framework/backward_test.cc]: remove unused code

File: paddle/framework/backward_test.cc [hunk 21]
```diff
@@ -0,0 +1,388 @@
...
+REGISTER_GRADIENT_OP(add, add_grad, f::EmptyOp);
+REGISTER_OP(fc, f::FcOp, f::FcOpMaker);
+REGISTER_OP(many_output_op, f::EmptyOp, f::ManyOutputOpMaker);
+REGISTER_GRADIENT_OP(many_output_op, many_output_op_grad, f::EmptyOp);
+
+TEST(Backward, simple_op_grad) {
+  auto fwd = f::OpRegistry::CreateOp("rowwise_add", {"X", "b"}, {"Out"}, {});
+  ASSERT_NE(fwd, nullptr);
+  auto gop = f::OpRegistry::CreateGradOp(*fwd);
+  ASSERT_EQ(1UL, gop->inputs_.size());
```
Comment (jacquesqiao) [inline paddle/framework/backward_test.cc]: 记得说GradientOp的Input是I,O,OG，这里为什么gop的input size为1？

Comment (reyoung) [inline paddle/framework/backward_test.cc]: 因为有一些Op的梯度可以不使用I和O，这个在注册的时候注册过了。

File: paddle/framework/backward_test.cc [hunk 22]
```diff
@@ -0,0 +1,388 @@
...
+TEST(Backward, simple_op_not_need_grad) {
+  auto fwd = f::OpRegistry::CreateOp("rowwise_add", {"X", "b"}, {"Out"}, {});
+  ASSERT_NE(fwd, nullptr);
+  auto gop = f::Backward(*fwd, {"X"});
+  ASSERT_EQ(std::find(gop->outputs_.begin(), gop->outputs_.end(),
+                      "X" + f::OperatorBase::GRAD_VAR_SUFFIX()),
+            gop->outputs_.end());
+
+  auto no_input_gop = f::Backward(*fwd, {"X", "b"});
+  ASSERT_NE(no_input_gop, nullptr);
```
Comment (jacquesqiao) [inline paddle/framework/backward_test.cc]: 好像没有做任何测试？

File: paddle/framework/backward.cc (hunk 18 above)
Comment (reyoung) [inline paddle/framework/backward.cc]: Because we do not export them to global symbols.

File: paddle/framework/backward.cc (hunk 17 above)
Comment (reyoung) [inline paddle/framework/backward.cc]: duplicated

File: paddle/operators/fill_zeros_like_op.h [hunk 23]
```diff
@@ -23,8 +23,8 @@ namespace operators {
 template <typename Place, typename T>
 class FillZerosLikeKernel : public framework::OpKernel {
 public:
-  void Compute(const framework::KernelContext& context) const override {
-    auto* output = context.Output(0)->GetMutable<framework::Tensor>();
+  void Compute(const framework::ExecutionContext& context) const override {
```
Comment (jacquesqiao) [inline paddle/operators/fill_zeros_like_op.h]: framework::xxx  ===> xxx

**Comment by jacquesqiao** [comment]
有一些注释的comment，也可以顺便update一下

File: paddle/framework/backward.cc (hunk 8 above)
Comment (reyoung) [inline paddle/framework/backward.cc]: It seems that `NOP` is only used in `Backward` and should not be used by users. So just write `NOP` here.

File: paddle/operators/fill_zeros_like_op.cu (hunk 10 above)
Comment (reyoung) [inline paddle/operators/fill_zeros_like_op.cu]: `zeros_like` is not a verb. But all operator should be a `verb`. I think fill_zeros_like is more explicit.

Also, `zeros_like` may be implemented without filling anything into memory but a flag. Here named `fill` indicate that actually a new memory is allocated.

File: paddle/operators/fill_zeros_like_op.h (hunk 11 above)
Comment (reyoung) [inline paddle/operators/fill_zeros_like_op.h]: We will add the unit test in other PR. @Canpio @dzhwinter

File: paddle/framework/operator.cc (hunk 12 above)
Comment (reyoung) [inline paddle/framework/operator.cc]: Done.

File: paddle/framework/backward.cc (hunk 14 above)
Comment (reyoung) [inline paddle/framework/backward.cc]: Done. BackwardRecursive.

File: paddle/framework/backward.cc (hunk 16 above)
Comment (reyoung) [inline paddle/framework/backward.cc]: done.

File: paddle/framework/backward.cc (hunk 19 above)
Comment (reyoung) [inline paddle/framework/backward.cc]: 是的。这个Op的实现不影响Backward算法的实现和单测。

File: paddle/framework/backward_test.cc (hunk 20 above)
Comment (reyoung) [inline paddle/framework/backward_test.cc]: Done.

File: paddle/framework/backward_test.cc (hunk 22 above)
Comment (reyoung) [inline paddle/framework/backward_test.cc]: Done.

File: paddle/operators/fill_zeros_like_op.h (hunk 23 above)
Comment (reyoung) [inline paddle/operators/fill_zeros_like_op.h]: Changed in Next PR

**Comment by reyoung** [comment]
@dzhwinter @Canpio @wangkuiyi 

I prepare to merge this PR soon, all code-style comments are followed. But it still lacks documentation and comments. But it has been a huge PR already. Documentation and comments will be added in following PR.

File: paddle/operators/fill_zeros_like_op.h (hunk 11 above)
Comment (dzhwinter) [inline paddle/operators/fill_zeros_like_op.h]: @reyoung get it

**Pull request merged**